### ISSUE TITLE
Caching issue for createApi when passing class as argument

### ISSUE THREAD

**Issue by tomekpaszek** [reduxjs/redux-toolkit#1767]
Hi!
I believe I've found an issue with the createApi, however my JS skills are too low to find the actual problem. 

The issue happens when I'm calling a query api (created with createApi) while passing class instance as an argument. Things seem to work fine as long as the query call succeeds. However, in case of a failure, React get stuck in a re-render loop trying to call the query repeatedly. The state jumps back and forth from `rejected` to `pending`.

The class I'm passing in as the argument is converted to a deterministic representation in a custom serializeQueryArgs callback. It seems like this is possibly not picked up by the cache in case on an error state is returned. 

I've tried reducing the problem to a simple use case:
```
import { createApi } from '@reduxjs/toolkit/query/react';

export class MyRequest {
  constructor(fail: boolean) {
    this.fail = fail;
  }
  fail: boolean;
  toObject = () => {
    return { fail: this.fail };
  };
}

export const testApi = createApi({
  reducerPath: 'testApi',
  baseQuery: async (arg, api, extraOptions) => {
    await new Promise((r) => setTimeout(r, 100));
    if (arg.fail)
      return {
        error: 'error',
      };
    return { data: 'ok' };
  },
  serializeQueryArgs: ({ queryArgs, endpointName }) => {
    return `${endpointName}(${JSON.stringify(queryArgs.toObject())})`;
  },
  endpoints: (builder) => ({
    test: builder.query<string, MyRequest>({
      query: (req) => req,
    }),
  }),
});

export const { useTestQuery } = testApi;
```

And then the API is called as such:
```
  const { data, error, isError, isLoading, isSuccess, isFetching } =
    useTestQuery(new MyRequest(true));
```

**Comment by phryneas** [comment]
You should really not put a class in there. Query args will be saved to the store, and the store should never contain class instances. Please pass serializable data in there - this is not something we support.

**Comment by tomekpaszek** [comment]
Hi @phryneas, thanks for the answer.
I am transforming the input class into an object in 'serializeQueryArgs'. Isn't that the the purpose of 'serializeQueryArgs'? It says in the docs 
> Accepts a custom function if you have a need to change the creation of cache keys for any reason.

Also, if i looked correctly the store store the key as a string and not an object and which is what i would expect.

**Comment by phryneas** [comment]
This is not about the cache key though. The originalArgs are stored in the store in case a query needs to be re-run, so your `query` function would need to be recalled. And there it suffers the same problems as every other unserializable object in Redux state: it cannot be rehydrated over SSR boundaries, it cannot be persisted with redux-persist, it might cause issues in the devtools.
So after a rehydrate in any way, a query could not be invalidated any more.

And it seems to cause whatever issue you have found here - probably the equality test in the hook. Your specific use case is *probably* solved in the current RTK 1.7 beta, but I'd still strongly advise against doing that.

**Comment by tomekpaszek** [comment]
Thank you for explanation. I will reconsider my approach. 
To give you some more background, I'm trying to wrap 'grpc-web' calls which take objects as arguments and I'm trying to avoid duplicate work.
